lib/transition/unpack.c

```diff
@@ -164,6 +164,130 @@ unpack_synapse(crm_graph_t * new_graph, xmlNode * xml_synapse)
     return new_synapse;
 }
 
+static xmlNode *
+get_start_action_xml(synapse_t *synapse)
+{
+    GListPtr lpc;
+
+    for (lpc = synapse->actions; lpc != NULL; lpc = lpc->next) {
+        crm_action_t *action = (crm_action_t *) lpc->data;
+        const char *operation = crm_element_value(action->xml, XML_LRM_ATTR_TASK);
+
+        if (action->type == action_type_rsc && safe_str_eq(operation, RSC_START)) {
+            return action->xml;
+        }
+    }
+
+    return NULL;
+}
+
+static int
+get_max_action_id(crm_graph_t *graph)
+{
+    GListPtr sIter;
+    int id = 0;
+
+    for (sIter = graph->synapses; sIter != NULL; sIter = sIter->next) {
+        synapse_t *synapse = (synapse_t *) sIter->data;
+        GListPtr aIter;
+
+        for (aIter = synapse->actions; aIter != NULL; aIter = aIter->next) {
+            crm_action_t *action = (crm_action_t *) aIter->data;
+
+            if (action->id > id) {
+                id = action->id;
+            }
+        }
+    }
+
+    return id;
+}
+
+static void destroy_action(crm_action_t * action);
+
+static gboolean
+inject_metadata(crm_graph_t *graph, synapse_t *parent_synapse)
+{
+    crm_action_t *new_input = NULL;
+    synapse_t *new_synapse = NULL;
+    xmlNode *synapse = NULL;
+    xmlNode *action_set = NULL;
+    xmlNode *op = NULL;
+    xmlNode *rsc = NULL;
+    xmlNode *attrs = NULL;
+    char *op_key = NULL;
+    xmlNode *start = NULL;
+    xmlNode *start_rsc = NULL;
+    xmlNode *start_attrs = NULL;
+
+    if (!graph || !parent_synapse) {
+        return FALSE;
+    }
+
+    start = get_start_action_xml(parent_synapse);
+    if (!start) {
+        return FALSE;
+    }
```

Comment:
I just mean that we need to know we have metadata before executing any action on a resource. Getting the metadata before a probe would cover the cases mentioned above, but there are other possibilities -- for example, someone can configure a recurring monitor with role=Stopped, even if they have probes disabled. An even less predictable situation (which I've never seen, but is supported in theory) is that an RA can define a custom action, and the user can configure it as a recurring action.

So, we need a generic solution, that would accomplish something like this:
* Go through each resource action in the graph
** If this is a start action, or we do not already have metadata cached for this agent:
*** If we have not yet injected a metadata call for this agent, inject a metadata call before this action.